packages/database/src/core/modelClasses/Entity.js

```diff
@@ -533,131 +529,92 @@ export class EntityModel extends MaterializedViewLogDatabaseModel {
-
-    return await this.runCachedFunction(cacheKey, async () => {
-      return await this.#getDescendantsRecursively(hierarchyId, entityIds, params);
-    });
-  }
-
-  async getAncestorsFromParentChildRelation(hierarchyId, entityIds, params = {}) {
-    const cacheKey = this.getCacheKey(this.getAncestorsFromParentChildRelation.name, arguments);
-
-    return await this.runCachedFunction(cacheKey, async () => {
-      return await this.#getAncestorsRecursively(hierarchyId, entityIds, params);
-    });
-  }
-
-  /**
-   * Recursively finds descendants using this.find and parent-child relations
-   * @param {string} hierarchyId - The hierarchy ID
-   * @param {string[]} parentIds - Array of parent entity IDs
-   * @param {object} params - Filter + Fields
-   * @param {EntityRecord[]} allDescendants - Accumulated descendants array
-   * @param {number} level - Current recursion level
-   * @returns {Promise<EntityRecord[]>} Array of descendant entity records
-   */
-  async #getDescendantsRecursively(
-    hierarchyId,
-    parentIds,
-    params = {},
-    allDescendants = [],
-    level = 0,
-  ) {
-    const { filter = {}, pageSize } = params;
-    if (!parentIds || parentIds.length === 0) {
-      return allDescendants;
+  async getEntitiesFromParentChildRelation(hierarchyId, entityIds, direction, params = {}) {
+    if (!entityIds || entityIds.length === 0) {
+      return [];
     }
 
-    const { generational_distance, ...restOfFilter } = filter;
+    const methodName =
+      direction === ENTITY_RELATION_TYPE.DESCENDANTS
+        ? this.getDescendantsFromParentChildRelation.name
+        : this.getAncestorsFromParentChildRelation.name;
 
-    if (generational_distance && level >= generational_distance) {
-      return allDescendants;
-    }
+    const cacheKey = this.getCacheKey(methodName, [hierarchyId, entityIds, direction, params]);
 
-    // Find direct children using parent-child relation
-    const children = await this.find(
-      {
-        ...restOfFilter,
-        [`${RECORDS.ENTITY_PARENT_CHILD_RELATION}.entity_hierarchy_id`]: hierarchyId,
-        [`${RECORDS.ENTITY_PARENT_CHILD_RELATION}.parent_id`]: parentIds,
-      },
-      {
-        joinWith: 'entity_parent_child_relation',
-        joinCondition: ['entity.id', 'child_id'],
-        limit: pageSize,
-      },
-    );
+    return await this.runCachedFunction(cacheKey, async () => {
+      const { filter = {}, fields, pageSize } = params;
+      const { generational_distance, ...restOfFilter } = filter;
 
-    const nextLevelParentIds = children.map(c => c.id);
+      const RECURSIVE_CTE_ALIAS = 'hierarchy';
 
-    // Recursively find children of children
-    if (nextLevelParentIds.length === 0) {
-      return allDescendants;
-    }
+      const results = await this.find(
+        {
+          [`${RECURSIVE_CTE_ALIAS}.generational_distance`]: generational_distance,
+          ...restOfFilter,
+        },
+        {
+          withRecursive: {
```

Comment:
This is the gist of the performance improvement. Previously, we fetch the entity relations via `entity_parent_child_relation` by recursively fetching next level in memory. That doesn't have very good performance when testing in real deployments. I've now changed it to do recursively in database level which improves the performance a lot